### PR TITLE
 9-back-end-endpoint - Database - Pakize Bozkurt

### LINKED ISSUE #9: Back-end endpoint
## Database endpoint/API

Call to DB returning a JSON objectr with users and bookings?
Dates?

## Acceptance Criteria

1. Returns:

users: {
  id:
  firstName:
  lastName:
  slack:
  email:
},

availability: {
  id:
....
  traineeId:
}

## Checklist:

TODO - update with Ender

- [ Y ] I have defined the user story
- [ Y ] I have defined some acceptance criteria
- [ Y ] I have included any implementation notes
- [ N ] I have included all relevant links

### PR DESCRIPTION
**Pull request by PakizeBozkurt**
Contained two tables, "trainees" and "availability", which are used to store information about trainees and their availability. The "trainees" table has four columns and two constraints, while the "availability" table has four columns and a foreign key constraint. The tables are well-designed and have appropriate constraints and relationships.
Endpoints:
The code uses the Express framework to handle HTTP requests and responses. It also uses the express-validator middleware to validate input data, and the cors middleware to enable Cross-Origin Resource Sharing (CORS). The jsonwebtoken library is used to generate and verify JWTs. The pg library is used to interact with the PostgreSQL database.

## Checklist:

- [ Y ] My code follows the style guidelines of this project
- [ Y ] I have carefully reviewed my own code
- [ Y ] I have commented my code
- [ Y ] I have updated any documentation

### CODE REVIEW

**Review by PakizeBozkurt — COMMENTED** [review]
The code is a Node.js backend application that provides APIs to register, login, get, create, update and delete availability records. It uses PostgreSQL as the database, bcrypt to hash passwords, and JSON Web Tokens (JWT) for authentication.

**Review by Nagehan1 — COMMENTED** [review]
Good job Pakize!

File: server/authEndpoint.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+	} catch (err) {
+		//eslint-disable-next-line
+		console.error(err);
+		res.status(500).json({ errors: [{ msg: "Server error" }] });
+	}
+});
+
+// Returns a list of all availability records in the database.
+// Only authenticated users can access this endpoint.
+app.get("/availabilities/", (req, res) => {
```
Comment (enderta) [inline server/authEndpoint.js]: const queryOptions = {
        daily: `WHERE availability_date = CURRENT_DATE`,
        weekly: `WHERE availability_date >= CURRENT_DATE - interval '7' day`,
        monthly: `WHERE date_trunc('month', availability_date) = date_trunc('month', CURRENT_DATE)`,
        name: `WHERE t.name = '${req.query.name}'`,
        email: `WHERE t.email = '${req.query.email}'`,
    };
    const queryFilter = queryOptions[req.query.filter] || '';
    const search = req.query.search || '';
    const query = `
        SELECT a.availability_date, a.topic, t.name, t.email
        FROM availability a
                 JOIN trainees t ON a.trainees_id = t.id ${queryFilter}
                AND (a.topic ILIKE '%${search}%' OR t.name ILIKE '%${search}%' OR t.email ILIKE '%${search}%')
        ORDER BY a.availability_date DESC
          
    `;

    const {rows} = await pool.query(query);
    res.json({
        data: rows,
    });

**Review by enderta — COMMENTED** [review]
can you please change the app.get("/availiblity") endpoint as I commented, in this new codes we can search and filter availabilities data easily without writing complex codes in our ui part

**Review by AL-ASHMALI — COMMENTED** [review]
It would be great to add more character to the password

File: server/authEndpoint.js (hunk 1 above)
Comment (PakizeBozkurt) [reply server/authEndpoint.js]: 👍🏻 Thank you Ender.

**Review by enderta — APPROVED** [review]

**Comment by PakizeBozkurt** [comment]
It's done!

**Review by enderta — APPROVED** [review]

**Pull request merged**